File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/model/orderstatus/TradeItOrderStatusParcelable.java
package it.trade.android.sdk.model.orderstatus;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

import it.trade.model.reponse.OrderLeg;
import it.trade.model.reponse.OrderStatusDetails;


public class TradeItOrderStatusParcelable implements Parcelable {

    private String groupOrderType;
    private String orderExpiration;
    private String orderType;
    private String groupOrderId;
    private List<TradeItOrderLegParcelable> orderLegs = new ArrayList<>();
    private String orderNumber;
    private String orderStatus;
    private List<TradeItOrderStatusParcelable> groupOrders = new ArrayList<>();

    public TradeItOrderStatusParcelable(OrderStatusDetails orderStatusDetails) {
        this.groupOrderType = orderStatusDetails.groupOrderType;
        this.orderExpiration = orderStatusDetails.orderExpiration;
        this.orderType = orderStatusDetails.orderType;
        this.groupOrderId = orderStatusDetails.groupOrderId;
        this.orderLegs = mapOrderLegToTradeItOrderLegParcelable(orderStatusDetails.orderLegs);
        this.orderNumber = orderStatusDetails.orderNumber;
        this.orderStatus = orderStatusDetails.orderStatus;
        this.groupOrders = mapOrderStatusDetailsToTradeItOrderStatusParcelable(orderStatusDetails.groupOrders);
    }

    public static List<TradeItOrderStatusParcelable> mapOrderStatusDetailsToTradeItOrderStatusParcelable(List<OrderStatusDetails> orderStatusDetailsList) {
        List<TradeItOrderStatusParcelable> orderStatusParcelableList = new ArrayList<>();
        for (OrderStatusDetails orderStatusDetails: orderStatusDetailsList) {
            orderStatusParcelableList.add(new TradeItOrderStatusParcelable(orderStatusDetails));
        }
        return orderStatusParcelableList;
    }

    public String getGroupOrderType() {
        return groupOrderType;
    }

    public String getOrderExpiration() {
        return orderExpiration;
    }

    public String getOrderType() {
        return orderType;
    }

    public String getGroupOrderId() {
        return groupOrderId;
    }

    public List<TradeItOrderLegParcelable> getOrderLegs() {
        return orderLegs;
    }

    public String getOrderNumber() {
        return orderNumber;
    }

    public String getOrderStatus() {
        return orderStatus;
    }

    public List<TradeItOrderStatusParcelable> getGroupOrders() {
        return groupOrders;
    }

    @Override
    public String toString() {
        return "TradeItOrderStatusParcelable{" +
                "groupOrderType='" + groupOrderType + '\'' +
                ", orderExpiration='" + orderExpiration + '\'' +
                ", orderType='" + orderType + '\'' +
                ", groupOrderId='" + groupOrderId + '\'' +
                ", orderLegs=" + orderLegs +
                ", orderNumber='" + orderNumber + '\'' +
                ", orderStatus='" + orderStatus + '\'' +
                ", groupOrders=" + groupOrders +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        TradeItOrderStatusParcelable that = (TradeItOrderStatusParcelable) o;

        if (groupOrderType != null ? !groupOrderType.equals(that.groupOrderType) : that.groupOrderType != null)
            return false;
        if (orderExpiration != null ? !orderExpiration.equals(that.orderExpiration) : that.orderExpiration != null)
            return false;
        if (orderType != null ? !orderType.equals(that.orderType) : that.orderType != null)
            return false;
        if (groupOrderId != null ? !groupOrderId.equals(that.groupOrderId) : that.groupOrderId != null)
            return false;
        if (!orderLegs.equals(that.orderLegs)) return false;
        if (orderNumber != null ? !orderNumber.equals(that.orderNumber) : that.orderNumber != null)
            return false;
        if (orderStatus != null ? !orderStatus.equals(that.orderStatus) : that.orderStatus != null)
            return false;
        return groupOrders.equals(that.groupOrders);

    }

    @Override
    public int hashCode() {
        int result = groupOrderType != null ? groupOrderType.hashCode() : 0;
        result = 31 * result + (orderExpiration != null ? orderExpiration.hashCode() : 0);
        result = 31 * result + (orderType != null ? orderType.hashCode() : 0);
        result = 31 * result + (groupOrderId != null ? groupOrderId.hashCode() : 0);
        result = 31 * result + orderLegs.hashCode();
        result = 31 * result + (orderNumber != null ? orderNumber.hashCode() : 0);
        result = 31 * result + (orderStatus != null ? orderStatus.hashCode() : 0);
        result = 31 * result + groupOrders.hashCode();
        return result;
    }

    private List<TradeItOrderLegParcelable> mapOrderLegToTradeItOrderLegParcelable(List<OrderLeg> orderLegs) {
        List<TradeItOrderLegParcelable> orderLegParcelables = new ArrayList<>();
        for (OrderLeg orderLeg: orderLegs) {
            orderLegParcelables.add(new TradeItOrderLegParcelable(orderLeg));
        }
        return orderLegParcelables;
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.groupOrderType);
        dest.writeString(this.orderExpiration);
        dest.writeString(this.orderType);
        dest.writeString(this.groupOrderId);
        dest.writeTypedList(this.orderLegs);
        dest.writeString(this.orderNumber);
        dest.writeString(this.orderStatus);
        dest.writeList(this.groupOrders);
    }

    protected TradeItOrderStatusParcelable(Parcel in) {
        this.groupOrderType = in.readString();
        this.orderExpiration = in.readString();
        this.orderType = in.readString();
        this.groupOrderId = in.readString();
        this.orderLegs = in.createTypedArrayList(TradeItOrderLegParcelable.CREATOR);
        this.orderNumber = in.readString();
        this.orderStatus = in.readString();
        this.groupOrders = new ArrayList<TradeItOrderStatusParcelable>();
        in.readList(this.groupOrders, TradeItOrderStatusParcelable.class.getClassLoader());
    }

    public static final Parcelable.Creator<TradeItOrderStatusParcelable> CREATOR = new Parcelable.Creator<TradeItOrderStatusParcelable>() {
        @Override
        public TradeItOrderStatusParcelable createFromParcel(Parcel source) {
            return new TradeItOrderStatusParcelable(source);
        }

        @Override
        public TradeItOrderStatusParcelable[] newArray(int size) {
            return new TradeItOrderStatusParcelable[size];
        }
    };
}

File: tradeit-android-sdk/src/main/kotlin/it/trade/android/sdk/model/orderstatus/TradeItOrderStatusParcelable.kt
package it.trade.android.sdk.model.orderstatus

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import it.trade.model.reponse.OrderLeg
import it.trade.model.reponse.OrderStatusDetails
import java.util.*


class TradeItOrderStatusParcelable : Parcelable {

    @SerializedName("groupOrderType")
    var groupOrderType: String? = null
        private set

    @SerializedName("orderExpiration")
    var orderExpiration: String? = null
        private set

    @SerializedName("orderType")
    var orderType: String? = null
        private set

    @SerializedName("groupOrderId")
    var groupOrderId: String? = null
        private set

    @SerializedName("orderLegs")
    var orderLegs: List<TradeItOrderLegParcelable>? = null
        private set

    @SerializedName("orderNumber")
    var orderNumber: String? = null
        private set

    @SerializedName("orderStatus")
    var orderStatus: String? = null
        private set

    @SerializedName("groupOrders")
    var groupOrders: List<TradeItOrderStatusParcelable> = ArrayList()
        private set

    constructor(orderStatusDetails: OrderStatusDetails) {
        this.groupOrderType = orderStatusDetails.groupOrderType
        this.orderExpiration = orderStatusDetails.orderExpiration
        this.orderType = orderStatusDetails.orderType
        this.groupOrderId = orderStatusDetails.groupOrderId
        this.orderLegs = mapOrderLegToTradeItOrderLegParcelable(orderStatusDetails.orderLegs)
        this.orderNumber = orderStatusDetails.orderNumber
        this.orderStatus = orderStatusDetails.orderStatus
        this.groupOrders = mapOrderStatusDetailsToTradeItOrderStatusParcelable(orderStatusDetails.groupOrders)
    }

    override fun toString(): String {
        return "TradeItOrderStatusParcelable{" +
                "groupOrderType='" + groupOrderType + '\''.toString() +
                ", orderExpiration='" + orderExpiration + '\''.toString() +
                ", orderType='" + orderType + '\''.toString() +
                ", groupOrderId='" + groupOrderId + '\''.toString() +
                ", orderLegs=" + orderLegs +
                ", orderNumber='" + orderNumber + '\''.toString() +
                ", orderStatus='" + orderStatus + '\''.toString() +
                ", groupOrders=" + groupOrders +
                '}'.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as TradeItOrderStatusParcelable?

        if (if (groupOrderType != null) groupOrderType != that!!.groupOrderType else that!!.groupOrderType != null)
            return false
        if (if (orderExpiration != null) orderExpiration != that.orderExpiration else that.orderExpiration != null)
            return false
        if (if (orderType != null) orderType != that.orderType else that.orderType != null)
            return false
        if (if (groupOrderId != null) groupOrderId != that.groupOrderId else that.groupOrderId != null)
            return false
        if (orderLegs != that.orderLegs) return false
        if (if (orderNumber != null) orderNumber != that.orderNumber else that.orderNumber != null)
            return false
        return if (if (orderStatus != null) orderStatus != that.orderStatus else that.orderStatus != null) false else groupOrders == that.groupOrders

    }

    override fun hashCode(): Int {
        var result = if (groupOrderType != null) groupOrderType!!.hashCode() else 0
        result = 31 * result + if (orderExpiration != null) orderExpiration!!.hashCode() else 0
        result = 31 * result + if (orderType != null) orderType!!.hashCode() else 0
        result = 31 * result + if (groupOrderId != null) groupOrderId!!.hashCode() else 0
        result = 31 * result + (orderLegs?.hashCode() ?: 0)
        result = 31 * result + if (orderNumber != null) orderNumber!!.hashCode() else 0
        result = 31 * result + if (orderStatus != null) orderStatus!!.hashCode() else 0
        result = 31 * result + groupOrders.hashCode()
        return result
    }

    private fun mapOrderLegToTradeItOrderLegParcelable(orderLegs: List<OrderLeg>): List<TradeItOrderLegParcelable> {
        val orderLegParcelables = ArrayList<TradeItOrderLegParcelable>()
        for (orderLeg in orderLegs) {
            orderLegParcelables.add(TradeItOrderLegParcelable(orderLeg))
        }
        return orderLegParcelables
    }


    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(this.groupOrderType)
        dest.writeString(this.orderExpiration)
        dest.writeString(this.orderType)
        dest.writeString(this.groupOrderId)
        dest.writeTypedList(this.orderLegs)
        dest.writeString(this.orderNumber)
        dest.writeString(this.orderStatus)
        dest.writeList(this.groupOrders)
    }

    protected constructor(`in`: Parcel) {
        this.groupOrderType = `in`.readString()
        this.orderExpiration = `in`.readString()
        this.orderType = `in`.readString()
        this.groupOrderId = `in`.readString()
        this.orderLegs = `in`.createTypedArrayList(TradeItOrderLegParcelable.CREATOR)
        this.orderNumber = `in`.readString()
        this.orderStatus = `in`.readString()
        this.groupOrders = ArrayList()
        `in`.readList(this.groupOrders, TradeItOrderStatusParcelable::class.java.getClassLoader())
    }

    companion object {

        fun mapOrderStatusDetailsToTradeItOrderStatusParcelable(orderStatusDetailsList: List<OrderStatusDetails>): List<TradeItOrderStatusParcelable> {
            val orderStatusParcelableList = ArrayList<TradeItOrderStatusParcelable>()
            for (orderStatusDetails in orderStatusDetailsList) {
                orderStatusParcelableList.add(TradeItOrderStatusParcelable(orderStatusDetails))
            }
            return orderStatusParcelableList
        }
        @JvmField
        val CREATOR: Parcelable.Creator<TradeItOrderStatusParcelable> = object : Parcelable.Creator<TradeItOrderStatusParcelable> {
            override fun createFromParcel(source: Parcel): TradeItOrderStatusParcelable {
                return TradeItOrderStatusParcelable(source)
            }

            override fun newArray(size: Int): Array<TradeItOrderStatusParcelable?> {
                return arrayOfNulls(size)
            }
        }
    }
}
